File: jvm/src/main/kotlin/com/jetbrains/signatureverifier/serialization/CMSAlgorithmProtectionAttributeInfo.kt
package com.jetbrains.signatureverifier.serialization

import kotlinx.serialization.Serializable
import org.bouncycastle.asn1.ASN1ObjectIdentifier
import org.bouncycastle.asn1.ASN1Sequence
import org.bouncycastle.asn1.DLSequence
import org.bouncycastle.asn1.DLTaggedObject
import org.bouncycastle.asn1.cms.Attribute
import org.bouncycastle.asn1.x509.AlgorithmIdentifier

@Serializable
data class CMSAlgorithmProtectionAttributeInfo(
  override val identifier: TextualInfo,
  val content: EncodableInfo
) : AttributeInfo() {

  constructor(attribute: Attribute) : this(
    TextualInfo.getInstance(attribute.attrType),
    attribute.attrValues.toEncodableInfo()
  )

  override fun getPrimitiveContent() = content.toPrimitive()
}

File: net/JetBrains.Serialization/src/Dataholders/CMSAlgorithmProtectionAttributeInfo.cs
using Newtonsoft.Json;
using Org.BouncyCastle.Asn1;
using Attribute = Org.BouncyCastle.Asn1.Cms.Attribute;

namespace JetBrains.Serialization;

[JsonObject(MemberSerialization.Fields)]
public class CMSAlgorithmProtectionAttributeInfo : AttributeInfo
{
  public override TextualInfo Identifier { get; }
  public IEncodableInfo Content { get; }

  public CMSAlgorithmProtectionAttributeInfo(TextualInfo identifier, IEncodableInfo content)
  {
    Identifier = identifier;
    Content = content;
  }

  public CMSAlgorithmProtectionAttributeInfo(Attribute attribute)
    : this(TextualInfo.GetInstance(attribute.AttrType), attribute.AttrValues.ToEncodableInfo())
  {
  }

  public override Asn1Encodable GetPrimitiveContent() => Content.ToPrimitive();
}